File: ojuz/ioi/2020/Day1/tickets_sol/solution/wy-subtask2.cpp
#include "tickets.h"
#include <bits/stdc++.h>
using namespace std;

vector<pair<int,int> > values[1505], diffs;
vector<int> fvalues;
int picked[1505];

long long find_maximum(int k, std::vector<std::vector<int>> d) {
	int c = d.size();
	int s = d[0].size();

	for (int i = 0; i < c; i++) {
		for (int j = 0; j < s; j++) values[i].push_back(make_pair(d[i][j], j));
		sort(values[i].begin(), values[i].end());
		diffs.push_back(make_pair(values[i].back().first + values[i][0].first, i));
	}
	sort(diffs.begin(), diffs.end());
	for (int i = 0; i < c; i++) {
		if (i < c/2) picked[diffs[i].second] = values[diffs[i].second][0].second;
		else picked[diffs[i].second] = values[diffs[i].second].back().second;
	}

	for (int i = 0; i < c; i++) fvalues.push_back(d[i][picked[i]]);
	sort(fvalues.begin(), fvalues.end());
	long long ans = 0;
	for (int i = 0; i < fvalues.size(); i++) ans += abs(fvalues[i] - fvalues[fvalues.size() / 2]);

	std::vector<std::vector<int>> answer;
	for (int i = 0; i < c; i++) {
		std::vector<int> row(s);
		for (int j = 0; j < s; j++) {
			if (j != picked[i]) row[j] = -1;
			else row[j] = 0;
		}
		answer.push_back(row);
	}
	allocate_tickets(answer);
	return ans;
}

File: ojuz/ioi/2020/Day1/tickets_sol/solution/wy-subtask2.java
import java.util.Arrays;

class tickets {
	long[] diffs;
	int[] picked;
	
	long combine(int x, int y) {
		return (((long)x << 32) | y);
	}
	int getx(long a) {
		return (int)(a>>32);
	}
	int gety(long a) {
		return (int)(a&((1<<31)-1));
	}
	
	long find_maximum(int k, int[][] d) {
		int c = d.length;
		int s = d[0].length;
		diffs = new long[c];
		picked = new int[c];
		
		for (int i = 0; i < c; i++) {
			diffs[i] = combine(d[i][0] + d[i][s-1], i);
		}
		Arrays.sort(diffs);
		for (int i = 0; i < c; i++) {
			if (i < c/2) {
				picked[gety(diffs[i])] = 0;
			} else {
				picked[gety(diffs[i])] = s-1;
			}
		}
		
		long ans = 0;
		for (int i = 0; i < c; i++) {
			ans += Math.abs(d[i][picked[i]] - getx(diffs[c/2])/2);
		}
		
		int[][] answer = new int[c][s];
		for(int i=0; i<c; i++) {
			for(int j=0; j<s; j++) {
				answer[i][j] = -1;
			}
			answer[i][picked[i]] = 0;
		}
		
		grader.allocate_tickets(answer);
		return ans;
	}
}
